dipy/reconst/dki.py

```diff
@@ -2296,29 +2411,48 @@ def ls_fit_dki(design_matrix, data, inverse_design_matrix,
     A = design_matrix
     y = np.log(data)
 
-    # DKI ordinary linear least square solution
-    result = np.dot(inverse_design_matrix, y)
+    if weights is not False:
+        # define W as "sqrt of the weights applied to *squared* residuals"
+        if type(weights) is np.ndarray:  # user supplied weights
+            W = np.diag(np.sqrt(weights))
+        else:  # Define weights as diag(fn) (fitted signal from DKI OLS)
+            result = np.dot(inverse_design_matrix, y)
+            W = np.diag(np.exp(np.dot(A, result)))
 
-    # Define weights as diag(yn**2)
```

Comment:
no that is not the case, **it depends on the meaning of W in weighted least squares**, and therefore on the way the linear algebra is implemented. You can either define the weights for Y or define them for Y^2. I have implemented the weights here to be for Y (or, as I may have poorly said # define W as "sqrt of the weights applied to *squared* residuals")

Note that previously in dki.py, the weights for `ls_fit_dki` were defined as diag(yn^2) due to the _previous_ implementation of the calculation, but in `cls_fit_dki` they were defined as diag(yn) due to the implementation there. But the mathematics of the weighting are the same in both cases.

Of course, I checked the the results of the new and old calculation (i.e. the parameters) were _exactly_ the same in both cases. Happy to check again, but I think you will find the new code is consistent with the old code.

Reply:
Ok got it! Can we add some comment line in the code mentioning that for future reference?

Reply:
I'll add a comment, and this got me thinking: best to add a test that implements WLS in a really obvious way, and checks that the WLS fit function actually gives the same result (no matter the implementation).

I can also test that the modified WLS fit functions in this PR give the same result if _provided_ with 'standard' WLS weights (since this PR provides the ability of WLS fit functions to accept external weights)